### PR TITLE
Metrics: categories, empty predictions, tests, fnmr@fmr

### PR DESCRIPTION
**Pull request by AlekseySh**
**CHANGELOG**

* Added support of empty predictions to the retrieval metrics. For example, it may be useful when we cut retrieval results by distance threshold).
* Moved categories handling to functional metrics from `EmbeddingMetrics` class, also updated `.md` example to show how to deal with categories.
* Added `calc_fnmr_at_fmr_rr`, removed `extract_pos_neg_dists` and `calc_fnmr_at_fmr_from_matrices`. Returned `fnmr@fmr` to `EmbeddingMetrics` (there was a todo).

TESTS
* Moved tests that use old formats of retrieval metrics to a separate folder: `...test_metrics/test_outdated/...`.
* Added a few new tests on retrieval metrics: test handling categories and empty predictions.
* Added test on `calc_fnmr_at_fmr_rr`.
* Added test that `EmbeddingMetrics` calculate all expected metrics.

### CODE REVIEW

File: oml/functional/knn.py [hunk 1]
```diff
@@ -53,7 +54,7 @@ def batched_knn(
     gt_ids = []
 
     # we do batching over first (queries) dimension
-    for i in range(0, nq, bs):
+    for i in tqdm(range(0, nq, bs), desc="Finding nearest neighbors."):
```
Comment (DaloroAT) [inline oml/functional/knn.py:57]: verbose?

File: oml/functional/metrics.py [hunk 2]
```diff
@@ -217,25 +239,34 @@ def calc_cmc(gt_tops: Sequence[BoolTensor], top_k: Tuple[int, ...]) -> List[Floa
...
+            return 0.0
+        else:
+            value = float(is_correct[:k_].any())
+            return value
 
     cmc = []
     for k in top_k:
-        cmc.append(FloatTensor([cmc_single(gts, k) for gts in gt_tops]))
+        cmc.append(
+            FloatTensor([cmc_single(gts, n_gt, k) for gts, n_gt in tqdm(zip(gt_tops, n_gts), desc="Computing CMC")])
```
Comment (DaloroAT) [inline oml/functional/metrics.py:263]: `desc=f"Computing CMC@{k}"`? 
"Computing CMC" better describes the outside loop.
The same comment for other functions.

File: oml/functional/metrics.py [hunk 3]
```diff
@@ -464,13 +520,10 @@ def calc_fnmr_at_fmr(pos_dist: Tensor, neg_dist: Tensor, fmr_vals: Tuple[float,
     return fnmr_at_fmr
 
 
-def calc_fnmr_at_fmr_from_matrices(
-    distance_matrix: FloatTensor, mask_gt: BoolTensor, fmr_vals: Tuple[float, ...]
-) -> TMetricsDict:
+def calc_fnmr_at_fmr_(pos_dist: FloatTensor, neg_dist: FloatTensor, fmr_vals: Tuple[float, ...]) -> TMetricsDict:
```
Comment (DaloroAT) [inline oml/functional/metrics.py:523]: Maybe something more meaningful than just a postfix `_` to the main `calc_fnmr_at_fmr` function? 😁

File: oml/retrieval/retrieval_results.py [hunk 4]
```diff
@@ -42,7 +42,7 @@ def __init__(
         for d, r in zip(distances, retrieved_ids):
             if not (d[:-1] <= d[1:]).all():
                 raise RuntimeError("Distances must be sorted.")
-            if not len(torch.unique(r)) == len(r):
+            if not len(torch.unique(r[:100])) == len(r[:100]):  # it's too expensive to check all the ids!
```
Comment (DaloroAT) [inline oml/retrieval/retrieval_results.py:45]: 99 🤔

File: oml/functional/metrics.py (hunk 3 above)
Comment (AlekseySh) [reply oml/functional/metrics.py:523]: okay
changed to `calc_fnmr_at_fmr_by_distances`

File: oml/retrieval/retrieval_results.py (hunk 4 above)
Comment (AlekseySh) [reply oml/retrieval/retrieval_results.py:47]: discussed offline

File: oml/functional/knn.py (hunk 1 above)
Comment (AlekseySh) [reply oml/functional/knn.py:57]: Having `verbose` parameter means adding this parameter to the all chain on function calls and to the config.
I don't think having this print may hurt... So, let's just keep it

**Comment by AlekseySh** [comment]
I also check that FNMR values computed on this branch and on the main brach are the same:
```
InShop: fnmr@fmr[0.1] = 0.2626
SOP: fnmr@fmr[0.1] = 0.1281
```

fyi @DaloroAT

**Pull request merged**